libgalois/include/katana/PropertyGraph.h

```diff
@@ -473,80 +487,200 @@ class KATANA_EXPORT PropertyGraph {
+  /// nullopt otherwise
+  /// (assumes that the node EntityTypeID exists)
+  std::optional<std::string> EdgeEntityTypeIDToAtomicTypeName(
+      EntityTypeID edge_entity_type_id) const {
+    auto& set_of_names =
+        edge_entity_type_id_to_type_names_.at(edge_entity_type_id);
+    if (set_of_names.size() == 1) {
+      return *set_of_names.cbegin();
+    }
+    return std::nullopt;
   }
 
-  /// \returns the number of node TypeSetIDs (including kUnknownType)
-  size_t GetNodeTypeSetIDsNum() const {
-    return node_type_set_id_to_type_names_.size();
+  /// \returns the set of node EntityTypeIDs that contain
+  /// the node type with @param name
+  /// (assumes that the node type exists)
+  const SetOfEntityTypeIDs& NodeTypeNameToEntityTypeIDs(
+      const std::string& name) const {
+    return node_type_name_to_entity_type_ids_.at(name);
   }
 
-  /// \returns the number of edge TypeSetIDs (including kUnknownType)
-  size_t GetEdgeTypeSetIDsNum() const {
-    return edge_type_set_id_to_type_names_.size();
+  /// \returns the set of edge EntityTypeIDs that contain
+  /// the edge type with @param name
+  /// (assumes that the edge type exists)
+  const SetOfEntityTypeIDs& EdgeTypeNameToEntityTypeIDs(
+      const std::string& name) const {
+    return edge_type_name_to_entity_type_ids_.at(name);
+  }
+
+  /// \returns the set of atomic node subtype names that intersect
+  /// the node EntityTypeID @param node_entity_type_id
+  /// (assumes that the node EntityTypeID exists)
+  const SetOfTypeNames& NodeEntityTypeIDToAtomicSubtypeNames(
+      EntityTypeID node_entity_type_id) const {
+    return node_entity_type_id_to_type_names_.at(node_entity_type_id);
+  }
+
+  /// \returns the set of atomic edge subtype names that intersect
+  /// the edge EntityTypeID @param edge_entity_type_id
+  /// (assumes that the edge EntityTypeID exists)
+  const SetOfTypeNames& EdgeEntityTypeIDToAtomicSubtypeNames(
+      EntityTypeID edge_entity_type_id) const {
+    return edge_entity_type_id_to_type_names_.at(edge_entity_type_id);
+  }
+
+  /// \returns the set of atomic node subtype names that intersect
+  /// the node EntityTypeID @param node_entity_type_id
+  /// (assumes that the node EntityTypeID exists)
+  TypeNamesRange NodeEntityTypeIDToAtomicSubtypeNamesRange(
+      EntityTypeID node_entity_type_id) const {
+    auto& type_names =
+        NodeEntityTypeIDToAtomicSubtypeNames(node_entity_type_id);
+    ;
+    return MakeStandardRange(type_names.cbegin(), type_names.cend());
+  }
+
+  /// \returns the set of atomic edge subtype names that intersect
+  /// the edge EntityTypeID @param edge_entity_type_id
+  /// (assumes that the edge EntityTypeID exists)
+  TypeNamesRange EdgeEntityTypeIDToAtomicSubtypeNamesRange(
+      EntityTypeID edge_entity_type_id) const {
+    auto& type_names =
+        EdgeEntityTypeIDToAtomicSubtypeNames(edge_entity_type_id);
+    ;
+    return MakeStandardRange(type_names.cbegin(), type_names.cend());
+  }
```

Comment:
Shouldn't these return ranges of `EntityTypeID` not string? Also why include "range" in the name? If they are just wrappers around `EdgeEntityTypeIDToAtomicSubtypeNames` I don't see their usefulness at all. You can always just iterate the `SetOfTypeNames`.

Reply:
I misunderstood your API requirements. 

Returning a range of entity-type-ids requires materializing a vector. I could do that but it's not used currently and when it is used, there will be overheads.

Reply:
I'm OK with returning an iterator range, but I'd like you to remove the variants that return Sets (above these 2 methods). Those shouldn't be needed given that these 2 methods return ranges.

Reply:
I don't see why you would need to materialize the vector to return a range. The range could be a filter or map operation over the internal data structures of `PropertyGraph`.

Reply:
I agree with Arthur that there should be functions to convert atomic-name and `EntityTypeID` but all other functions should be in terms of `EntityTypeID`. Returning a set of `EntityTypeID` is required to do set operations - range won't suffice for that. Returning a range requires materialization because `std::bitset` does not provide an iterator for set bits. Arthur, I did not understand what you meant by filter or map?

Reply:
By filter or map, I just meant we could create a range/iterators that computes the values needed during iteration, eliminating the need to materialize. I said filter/map because that's how you express these kinds of lazy collection operations in many languages. C++ doesn't actually have such operations, but the concepts still exist and can be implemented.

Just because std::bitset doesn't provide such an iterator doesn't mean we cannot.

One thing we could do is have a special collection class that uses a bitset representation internally and provides the required iterators and a way to materialize the set or bitset when required.

Reply:
I think I now understand. I had implemented such an iterator for `DynamicBitset`. Maybe we can repurpose that iterator for this.

Reply:
All this said. I'm gonna back off on this. I don't want to create too much work without much gain. C++ makes good abstraction so difficult that it may not be worth it. Especially given that the idiomatic style of C++ tends to be very poorly abstracted.

Reply:
Thanks both of you for all the comments/suggestions. I'll handle this soon.